### PR TITLE
use span<> to return the block_info for a piece

### PR DESCRIPTION
**Pull request by arvidn**
to improve range checks and loops

### CODE REVIEW

File: include/libtorrent/span.hpp [hunk 1]
```diff
@@ -105,12 +105,19 @@ namespace libtorrent
...
 		}
 
-		T& operator[](std::size_t const idx)
+		T& operator[](std::size_t const idx) const
 		{
 			TORRENT_ASSERT(idx < m_len);
 			return m_ptr[idx];
 		}
 
+		T& operator[](int const idx) const
```
Comment (aldenml) [inline include/libtorrent/span.hpp]: I didn't review completely the PR, but I see `span<T>` like a very interoperable type, similar to `std::vector<T>`, and that means classic `std::size_t` index. Maybe you can use `typed_span<T, I>` in the place you need to index by `int`, similar to `aux::vector`.

Comment (arvidn) [inline include/libtorrent/span.hpp]: yeah, I thought about that a bit. you mean that ``span`` should be ready to be replaced by gsl or some other standard type? so this utility feature may get in the way. I agree with that, they should be sepate, I just wasn't sure it was worth separating them up-front, rather then put it off until it's actually needed

Comment (aldenml) [inline include/libtorrent/span.hpp]: yes, and `aux::typed_span` is already there

Comment (arvidn) [inline include/libtorrent/span.hpp]: oh, I see

Comment (aldenml) [inline include/libtorrent/span.hpp]: looking at it, I believe I missed the `IndexType` in the return type of the methods.

File: src/torrent.cpp [hunk 2]
```diff
@@ -1855,30 +1855,25 @@ namespace libtorrent
 			{
 				std::vector<std::shared_ptr<torrent>> ts = m_ses.find_collection(c);
 
-				for (std::vector<std::shared_ptr<torrent>>::iterator k = ts.begin()
-					, end2(ts.end()); k != end2; ++k)
+				for (auto const& t : ts)
```
Comment (aldenml) [inline src/torrent.cpp]: sometimes I wonder how exactly lists of `std::shared_ptr` should be iterated, by ref, or by copy (just a comment)

Comment (arvidn) [inline src/torrent.cpp]: by reference. there's no need to incur the cost of atomic increment and decrement for the loop

Comment (aldenml) [inline src/torrent.cpp]: thanks!

**Pull request merged**